src/mlpack/methods/cf/normalization/user_mean_normalization.hpp

```diff
@@ -0,0 +1,157 @@
+/**
+ * @file user_mean_normalization.hpp
+ * @author Wenhao Huang
+ *
+ * This class performs user mean normalization on raw ratings. In another
+ * word, this class is used to remove global effect of user mean.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_CF_NORMALIZATION_USER_MEAN_NORMALIZATION_HPP
+#define MLPACK_METHODS_CF_NORMALIZATION_USER_MEAN_NORMALIZATION_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace cf {
+
+/**
+ * This normalization class performs user mean normalization on raw ratings.
+ *
+ * An example of how to use UserMeanNormalization in CF is shown below:
+ *
+ * @code
+ * extern arma::mat data; // (user, item, rating) table
+ * extern arma::Col<size_t> users; // users seeking recommendations
+ * arma::Mat<size_t> recommendations; // Recommendations
+ *
+ * CFType<UserMeanNormalization> cf(data);
+ *
+ * // Generate 10 recommendations for all users.
+ * cf.GetRecommendations(10, recommendations);
+ * @endcode
+ */
+class UserMeanNormalization
+{
+ public:
+  // Empty constructor.
+  UserMeanNormalization() { }
+
+  /**
+   * Normalize the data by subtracting user mean from each of existing ratings.
+   *
+   * @param data Input dataset in the form of coordinate list.
+   */
+  void Normalize(arma::mat& data)
+  {
+    const size_t userNum = arma::max(data.row(0)) + 1;
+    userMean = arma::vec(userNum, arma::fill::zeros);
+    // Number of ratings for each user.
+    arma::Row<size_t> ratingNum(userNum, arma::fill::zeros);
+
+    // Sum ratings for each user.
+    data.each_col([&](arma::vec& datapoint) {
+      const size_t user = (size_t) datapoint(0);
+      const double rating = datapoint(2);
+      userMean(user) += rating;
+      ratingNum(user) += 1;
+    });
+
+    // Calculate user mean and subtract user mean from ratings.
+    // Set user mean to 0 if the user has no rating.
+    for (size_t i = 0; i < userNum; i++)
+    {
+      if (ratingNum(i) != 0)
+        userMean(i) /= ratingNum(i);
+    }
+
+    data.each_col([&](arma::vec& datapoint) {
+      const size_t user = (size_t) datapoint(0);
+      datapoint(2) -= userMean(user);
+      // The algorithm omits rating of zero. If normalized rating equals zero,
+      // it is set to the smallest positive double value.
+      if (datapoint(2) == 0)
+        datapoint(2) = std::numeric_limits<double>::min();
+    });
+  }
+
+  /**
+   * Normalize the data by subtracting user mean from each of existing rating.
+   *
+   * @param cleanedData Sparse matrix data.
```

Comment:
Perhaps it's a good idea to clarify the comment somewhat; what do you think about: ``Input dataset as a sparse matrix. ``